applications/ShapeOptimizationApplication/python_scripts/algorithm_trust_region.py

```diff
@@ -0,0 +1,682 @@
+                nodal_variable_mapped = KratosGlobals.GetVariable("DC"+str(itr+1)+"DX_MAPPED")
+                self.mapper.MapToDesignSpace(nodal_variable, nodal_variable_mapped)
+                self.mapper.MapToGeometrySpace(nodal_variable_mapped, nodal_variable_mapped)
+
+                # Damping
+                if self.is_damping_specified:
+                    self.damping_utilities.DampNodalVariable(nodal_variable_mapped)
+
+            # Process inequality constraints
+            else:
+                ineq_id = con["identifier"].GetString()
+
+                ineq_gradients_dict = self.communicator.getStandardizedGradient(ineq_id)
+
+                nodal_variable = KratosGlobals.GetVariable("DC"+str(itr+1)+"DX")
+                WriteDictionaryDataOnNodalVariable(ineq_gradients_dict, self.optimization_model_part, nodal_variable)
+
+                # Projection on surface normals
+                if con["project_gradient_on_surface_normals"].GetBool():
+                    self.geometry_utilities.ProjectNodalVariableOnUnitSurfaceNormals(nodal_variable)
+
+                # Damping
+                if self.is_damping_specified:
+                    self.damping_utilities.DampNodalVariable(nodal_variable)
+
+                # Mapping
+                nodal_variable_mapped = KratosGlobals.GetVariable("DC"+str(itr+1)+"DX_MAPPED")
+                self.mapper.MapToDesignSpace(nodal_variable, nodal_variable_mapped)
+                self.mapper.MapToGeometrySpace(nodal_variable_mapped, nodal_variable_mapped)
+
+                # Damping
+                if self.is_damping_specified:
+                    self.damping_utilities.DampNodalVariable(nodal_variable_mapped)
+
+    # --------------------------------------------------------------------------
+    def __ConvertAnalysisResultsToLengthDirectionFormat(self):
+        # Convert objective results
+        obj = self.specified_objectives[0]
+        obj_id = obj["identifier"].GetString()
+
+        nodal_variable = KratosGlobals.GetVariable("DF1DX")
+        nodal_variable_mapped = KratosGlobals.GetVariable("DF1DX_MAPPED")
+
+        obj_value = self.communicator.getStandardizedValue(obj_id)
+        obj_gradient = ReadNodalVariableToList(self.design_surface, nodal_variable)
+        obj_gradient_mapped = ReadNodalVariableToList(self.design_surface, nodal_variable_mapped)
+
+        dir_obj, len_obj = self.__ConvertToLengthDirectionFormat(obj_value, obj_gradient, obj_gradient_mapped)
+        dir_obj = dir_obj
+
+        # Convert constraints
+        len_eqs = []
+        dir_eqs = []
+        len_ineqs = []
+        dir_ineqs = []
+
+        for itr in range(self.specified_constraints.size()):
+            con = self.specified_constraints[itr]
+
+            # Convert equality constraints
+            if con["type"].GetString()=="=":
+                eq_id = con["identifier"].GetString()
+
+                nodal_variable = KratosGlobals.GetVariable("DC"+str(itr+1)+"DX")
```

Comment:
a general function to create the variable name from the index of a constraint/objective could hide these string operations that exist all over the place.